Implement torch.squareform

lhuang04 opened the pull request:
Summary:
implement the equivalent of
https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.distance.squareform.html#scipy.spatial.distance.squareform
as torch.squareform.

Differential Revision: D13568360

Test Plan:
# Basic functional test
python test/test_torch.py TestTorch.test_squareform 
# Autograd
python test/test_autograd.py  TestAutograd.test_squareform_1d TestAutograd.test_squareform_2d TestAutograd.test_squareform_1d_large TestAutograd.test_squareform_2d_large




# Benchmark in ipython
import torch
from scipy.spatial.distance import squareform, pdist
cuda = torch.device('cuda')
#1024x1024
v523776 = torch.arange(1, 523777, dtype=torch.double)
#2048x2048
v2096128 = torch.arange(1, 2096129, dtype=torch.double)
#10000x10000
v49995000 = torch.arange(1, 49995001, dtype=torch.double)

c523776 = torch.arange(1, 523777, dtype=torch.double, device = cuda)
c2096128 = torch.arange(1, 2096129, dtype=torch.double, device = cuda)
c49995000 = torch.arange(1, 49995001, dtype=torch.double, device = cuda)
lst = [v523776, v2096128, v49995000]
clst = [c523776, c2096128, c49995000]

def foo(lst, f):
  for vec in lst:
    ysq = f(vec)
    yvec = f(ysq)


timeit foo(lst, squareform)

timeit foo(lst, torch.squareform)

timeit foo(clst, torch.squareform)

# scipy
In [5]: timeit foo(lst, squareform)
3.62 s ± 28.6 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
# torch cpu
In [6]: timeit foo(lst, torch.squareform)
2.11 s ± 49.3 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
# torch cuda
In [7]: timeit foo(clst, torch.squareform)
41.6 ms ± 11.3 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

gchanan commented on torch/nn/functional.py:
```diff
@@ -2783,6 +2783,36 @@ def pairwise_distance(x1, x2, p=2., eps=1e-6, keepdim=False):
+    tensor([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10])
+
+    """
+    assert input.dim() == 1 or input.dim() == 2, 'dimension of the input need be 1 or 2'
```
squareform should be a function at the C++ (native_function) level.  The downside of doing it this way is that the C++ API does not match the python API.

gchanan commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -1261,6 +1261,18 @@
   variants: function
 
+- func: to_squareform(Tensor self) -> Tensor
+  python_module: nn
```
these shouldn't go int the nn module, this function doesn't really have anything to do with neural networks specifically.

gchanan commented on aten/src/ATen/native/Distance.cpp:
```diff
@@ -75,4 +75,82 @@ Tensor cosine_similarity(const Tensor& x1, const Tensor& x2, int64_t dim, double
+Tensor to_squareform(const Tensor& self) 
+{
+  AT_CHECK(self.dim() == 1,
+    "expected vector-form distance as input, dim as 1, but got ", self.dim());
```
this error message is a little awkward.  How about something like "Expected 1-dimensional (vector-form) distance, but got tensor of dimensionality "...

gchanan commented on aten/src/ATen/native/Distance.cpp:
```diff
@@ -75,4 +75,82 @@ Tensor cosine_similarity(const Tensor& x1, const Tensor& x2, int64_t dim, double
+  AT_CHECK(self.size(0) == self.size(1),
+    "expected square-form distance as input, size(0) ",
+    self.size(0), " != size(1) ", self.size(1));
+  int64_t d = self.size(0), size = d * (d - 1) / 2;
```
can you put these on separate lines?

gchanan commented on aten/src/ATen/native/Distance.cpp:
```diff
@@ -75,4 +75,82 @@ Tensor cosine_similarity(const Tensor& x1, const Tensor& x2, int64_t dim, double
+{
+  AT_CHECK(self.dim() == 1,
+    "expected vector-form distance as input, dim as 1, but got ", self.dim());
+  int64_t size = self.size(0), d = (std::sqrt(8 * size + 1) + 1) / 2;
```
nit: can you put these on separate lines?

gchanan replied:
can you add a comment explaining the "d" calculation?

gchanan commented on aten/src/ATen/native/Distance.cpp:
```diff
@@ -75,4 +75,82 @@ Tensor cosine_similarity(const Tensor& x1, const Tensor& x2, int64_t dim, double
+  {
+    for (int64_t j = i + 1; j < d && idx < size; j++)
+    {
+      out_tensor[i][j] = self[idx];
```
I think this will be slow because you create multiple tensors each time through the loop (can you benchmark against numpy?).  Instead you could operate on .data and use the strides.

fmassa commented:
Question: why not dispatching to the `index` / `index_put_` kernels with the right indexing elements? This would involve some extra iterations maybe (generating the indices), but it would support GPU and backward computation off the box

ezyang commented:
@fmassa I suggested to @lhuang04 to implement it directly, mostly because it didn't seem wise to have to generate another tensor (order size of the output) to setup the indices, when it's completely avoidable. Also `index_put_` won't be differentiable ;)

lhuang04 commented:
so far the benchmark status is like this,
#scipy squareform
>>> timeit("foo(lst, squareform)", globals=globals(), number=100)
0.02014931500889361
#current CPU implementation
>>> timeit("foo(lst, torch.squareform)", globals=globals(), number=100)
0.01594047201797366
#current GPU implementation(still in progress)
>>> timeit("foo(clst, torch.squareform)", globals=globals(), number=100)
2.789320666110143

gchanan commented on aten/src/ATen/native/Distance.cpp:
```diff
@@ -75,4 +88,269 @@ Tensor cosine_similarity(const Tensor& x1, const Tensor& x2, int64_t dim, double
+{
+  AT_CHECK((self.dim() == 1 || self.dim() == 2),
+    "expected vector-form distance or square-form distance as input",
+    " but the dimension got from input is ", self.dim());
```
nit: dimension of input is.

gchanan commented on aten/src/ATen/native/Distance.cpp:
```diff
@@ -75,4 +88,269 @@ Tensor cosine_similarity(const Tensor& x1, const Tensor& x2, int64_t dim, double
+      to_squareform_frame<scalar_t> (self.data<scalar_t>(),
+        out_tensor.data<scalar_t>(), d, size);
+      }
+      );
```
nit: spacing is wrong here.

gchanan commented on aten/src/ATen/native/Distance.cpp:
```diff
@@ -75,4 +88,269 @@ Tensor cosine_similarity(const Tensor& x1, const Tensor& x2, int64_t dim, double
+static void to_squareform_frame(
+    scalar_t *input_p, scalar_t *output_p, int64_t dimension, int64_t size)
+{
+  for (int64_t i = 0, idx = 0; i < dimension; i++)
```
this won't work if the input isn't contiguous (add a test).

gchanan commented on aten/src/ATen/native/Distance.cpp:
```diff
@@ -75,4 +88,269 @@ Tensor cosine_similarity(const Tensor& x1, const Tensor& x2, int64_t dim, double
 
+template <typename scalar_t>
+static void to_squareform_frame(
+    scalar_t *input_p, scalar_t *output_p, int64_t dimension, int64_t size)
```
you are using terms in non-standard ways which makes this a little difficult to read.  dimension usually refers to an dimension index and size usually refers to the size of the dimension, but these are both referring to sizes here.  How about something like "square_dim_size" and "vector_size" or something?

gchanan commented on aten/src/ATen/native/Distance.cpp:
```diff
@@ -75,4 +88,269 @@ Tensor cosine_similarity(const Tensor& x1, const Tensor& x2, int64_t dim, double
+{
+  for (int64_t i = 0, idx = 0; i < dimension; i++)
+  {
+    for (int64_t j = i + 1; j < dimension && idx < size; j++)
```
do you need the idx < size?

gchanan commented on aten/src/ATen/native/Distance.cpp:
```diff
@@ -75,4 +88,269 @@ Tensor cosine_similarity(const Tensor& x1, const Tensor& x2, int64_t dim, double
 }
 
+template <typename scalar_t>
+static void to_squareform_frame(
```
can you put all the to_squareform and from_squareform things together, respectively?  This is unnecessarily difficult to read.

gchanan commented on aten/src/ATen/native/Distance.cpp:
```diff
@@ -75,4 +88,269 @@ Tensor cosine_similarity(const Tensor& x1, const Tensor& x2, int64_t dim, double
+    "Expected 2-dimensional (square-form) distance,",
+    " but got tensor of dimensionality ", self.dim());
+  AT_CHECK(self.size(0) == self.size(1),
+    "expected square-form distance as input, size(0) ",
```
this isn't a particularly helpful error message; the user might not know what a square-form distance is.  What about just saying square matrix?

gchanan commented on aten/src/ATen/native/Distance.cpp:
```diff
@@ -75,4 +88,269 @@ Tensor cosine_similarity(const Tensor& x1, const Tensor& x2, int64_t dim, double
+    " but got tensor of dimensionality ", self.dim());
+  int64_t size = self.size(0);
+  int64_t d = getSquareFormDimension(size);
+  Tensor out_tensor = at::zeros({d, d}, self.options());
```
it's wasteful to set all the values to 0 when you only need the diagonals to be 0.

gchanan commented on aten/src/ATen/native/Distance.cpp:
```diff
@@ -75,4 +88,269 @@ Tensor cosine_similarity(const Tensor& x1, const Tensor& x2, int64_t dim, double
+  AT_CHECK(grad.dim() == 2,
+    "Expected 2D grad, but the dimension of the grad is ", grad.dim());
+  AT_CHECK(grad.size(0) == grad.size(1),
+    "expected square-form distance as input, size(0) ",
```
same here.

gchanan commented on aten/src/ATen/native/Distance.cpp:
```diff
@@ -75,4 +88,269 @@ Tensor cosine_similarity(const Tensor& x1, const Tensor& x2, int64_t dim, double
+      to_squareform_backward_frame<scalar_t> (grad.data<scalar_t>(),
+        out_tensor.data<scalar_t>(), d, size);
+      }
+      );
```
spacing.

gchanan commented on test/test_torch.py:
```diff
@@ -1140,6 +1140,29 @@ def test_pdist_norm(self):
+              dtype=torch.double,
+              requires_grad=True
+            )
+            self.assertTrue(torch.autograd.gradcheck(lambda x: torch.squareform(x), (vec)))
```
don't do gradchecks in here, do it in test_autograd.

gchanan commented on aten/src/ATen/native/Distance.cpp:
```diff
@@ -75,4 +88,269 @@ Tensor cosine_similarity(const Tensor& x1, const Tensor& x2, int64_t dim, double
+
+Tensor to_squareform_backward_cuda(const Tensor& grad, const Tensor& self)
+{
+  AT_CHECK(grad.dim() == 2,
```
this cuda stuff shouldn't be here, this isn't a cuda implementation.  just check if the input is cuda and throw an error (or put in a dispatch only for CPU).

gchanan commented on test/test_torch.py:
```diff
@@ -1140,6 +1140,29 @@ def test_pdist_norm(self):
+    def test_squareform(self):
+        devices = ['cpu'] if not torch.cuda.is_available() else ['cpu', 'cuda']
+        for device in devices:
+            vec = torch.arange(1, 7, device=device)
```
how about edge cases, i.e. d == 0, d == 1, (0,0), (1,1), error cases?

gchanan commented on aten/src/ATen/native/SquareForm.cpp:
```diff
@@ -0,0 +1,181 @@
+static void to_squareform_frame(
+    scalar_t *input_p, scalar_t *output_p, int64_t square_dim_size, int64_t vector_size)
+{
+  for (int64_t i = 0, idx = 0; i < square_dim_size; i++)
```
you should write a cuda kernel, see the other examples in this directory.

lhuang04 commented:
# Benchmark in ipython
import torch
from scipy.spatial.distance import squareform, pdist
cuda = torch.device('cuda')
#1024x1024
v523776 = torch.arange(1, 523777, dtype=torch.double)
#2048x2048
v2096128 = torch.arange(1, 2096129, dtype=torch.double)
#10000x10000
v49995000 = torch.arange(1, 49995001, dtype=torch.double)

c523776 = torch.arange(1, 523777, dtype=torch.double, device = cuda)
c2096128 = torch.arange(1, 2096129, dtype=torch.double, device = cuda)
c49995000 = torch.arange(1, 49995001, dtype=torch.double, device = cuda)
lst = [v523776, v2096128, v49995000]
clst = [c523776, c2096128, c49995000]

def foo(lst, f):
  for vec in lst:
    ysq = f(vec)
    yvec = f(ysq)


timeit foo(lst, squareform)

timeit foo(lst, torch.squareform)

timeit foo(clst, torch.squareform)

# scipy
In [5]: timeit foo(lst, squareform)
3.62 s ± 28.6 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
# torch cpu
In [6]: timeit foo(lst, torch.squareform)
2.11 s ± 49.3 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
# torch cuda
In [7]: timeit foo(clst, torch.squareform)
41.6 ms ± 11.3 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

gchanan commented on aten/src/ATen/native/SquareForm.cpp:
```diff
@@ -0,0 +1,168 @@
+#include "ATen/ATen.h"
```
use <> instead of ""

gchanan commented on test/test_autograd.py:
```diff
@@ -2831,21 +2831,23 @@ def check(name):
+                is_method = hasattr(self_variable, name)
+                if is_method:
+                    output_variable = getattr(self_variable, name)(*args_variable, **kwargs_variable)
+                    if not exclude_tensor_method(name, test_name):
```
is exclude_tensor_method here not sufficient?

lhuang04 replied:
no, it is not sufficient, the line above it (2834) will fail if the name is not a tensor method.

lhuang04 replied:
2836 in the new code

gchanan replied:
FYI you can link the code directly, click "view file" under "Files Changed" such as 
`https://github.com/pytorch/pytorch/blob/c85170bd2d0bc09c4bc91c621b96345b01069f16/test/test_autograd.py#L2836`

which is displayed as:
https://github.com/pytorch/pytorch/blob/c85170bd2d0bc09c4bc91c621b96345b01069f16/test/test_autograd.py#L2836.

gchanan commented on test/test_torch.py:
```diff
@@ -1140,6 +1140,45 @@ def test_pdist_norm(self):
+            self.assertEqual(vec, torch.squareform(torch.squareform(vec)))
+
+            # test contiguous
+            non_contiguous_vec = torch.tensor([[0, 1], [0, 2], [0, 3], [0, 4], [0, 5], [0, 6]])[:, 1]
```
assert it isn't contiguous.

gchanan commented on test/test_torch.py:
```diff
@@ -1140,6 +1140,45 @@ def test_pdist_norm(self):
+
+            # expected failure case
+            vec = torch.arange(1, 58, device=device)
+            self.assertRaises(RuntimeError, lambda: torch.squareform(vec))
```
can you check the message?

gchanan commented on test/common_methods_invocations.py:
```diff
@@ -1023,6 +1031,10 @@ def unpack_variables(args):
     'test_slogdet_1x1_neg_det',
     'test_slogdet_neg_det',
     'test_slogdet_symmetric',
+    'test_squareform_1d',
```
why are these included here?

gchanan commented on test/common_methods_invocations.py:
```diff
@@ -797,6 +797,14 @@ def method_tests():
 
+def torch_function_tests():
+    return [
+        ('squareform', (1,), NO_ARGS, '1d'),
```
the point of this file is to automatically run these tests through different functionality, i.e. autograd, jit, etc.  This won't be run through the JIT tests -- does the JIT work if you add these tests to the above list?

gchanan commented on test/test_autograd.py:
```diff
@@ -2831,21 +2831,23 @@ def check(name):
+                        return output_process_fn(output)
+
+                    if not is_inplace and name not in EXCLUDE_GRADCHECK:
+                        run_grad_and_gradgrad_checks(self, name, test_name, fn,
```
I don't really get what the existing tests are doing that exclude_tensor_method but still go here.  Can you explain?

gchanan commented on test/test_torch.py:
```diff
@@ -1139,6 +1139,57 @@ def test_pdist_norm(self):
+            sq = [[0]]
+            full_square = torch.tensor(sq, device=device)
+            full_square_ret = torch.squareform(vec)
+            self.assertEqual(full_square_ret, full_square, "{} did not equal expected result".format(full_square_ret))
```
shouldn't the default message be something like you added here?

gchanan commented on test/test_torch.py:
```diff
@@ -1139,6 +1139,57 @@ def test_pdist_norm(self):
+            self.assertRaises(RuntimeError, lambda: torch.squareform(rect))
+            # test error message
+            try:
+                torch.squareform(vec)
```
I think assertRaisesRegex is what you want.

gchanan commented on test/test_torch.py:
```diff
@@ -1139,6 +1139,57 @@ def test_pdist_norm(self):
+            # smallest case [1]
+            vec = torch.arange(1, 2, device=device)
+            sq = [[0, 1], [1, 0]]
+            full_square = torch.tensor(sq, device=device)
```
what about tests for valid sizes, but invalid inputs, e.g. non-symmetric matrices?

lhuang04 committed:
Implement torch.squareform (#15679)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/15679

implement the equivalent of
https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.distance.squareform.html#scipy.spatial.distance.squareform
as torch.squareform.

Differential Revision: D13568360

fbshipit-source-id: ab2f28ce7f28f5dc09e7f53f3c4e21704e1882b9

LemonPi commented:
This would be nice to have for the current torch.pdist function; is this still being worked on?

TobiasFischerP4D commented:
I am wondering the same thing. Is this under active development? Looks like it is quite close to being finished!

ezyang commented:
Nope, this diff has been abandoned. If someone wants to pick it up and drive it through, that would be great!
